### PR TITLE
[FLINK-5224] [table] Improve UDTF: emit rows directly instead of buffering them

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR is an improvement for user-defined table function in Flink TableAPI & SQL. I didn't add new tests because the existing tests are sufficient.

The core design is code generating a `Collector` and register it into the instance of `TableFunction`, and emit the rows generated by the UDTF directly instead of buffering them.

### CODE REVIEW

**Comment by twalthr** [comment]
I will shepherd this PR.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -326,6 +346,52 @@ class CodeGenerator(
...
+    *             valid Java class identifier.
+    * @param bodyCode body code for the collector method
+    * @param returnType The type information of the element collected by the collector
+    * @return instance of GeneratedFunction
+    */
+  def generateTableFunctionCollector(
+    name: String,
+    bodyCode: String,
+    returnType: TypeInformation[Any])
+  : GeneratedFunction[TableFunctionCollector[Any]] = {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: I would introduce a new `GeneratedCollector` class as this is actually not a Flink `Function`.

**Comment by wuchong** [comment]
Thank you for your reviewing @twalthr , I will update this PR ASAP.

**Comment by wuchong** [comment]
Hi @twalthr , I have updated this PR. 

I refactored the code of `FlinkCorrelate#generateFunction` to separate the generation of `FlatMapFunction` and `TableFunctionCollector`. This can fix the problem that the generated `TableFunctionCollector` contains another `TableFunctionCollector` instance variable and makes the logic clean.

**Comment by twalthr** [comment]
Thanks for updating the PR. Looks very good now. I will merge this.

**Pull request closed without merging**